### PR TITLE
change interface between FNS and worker

### LINKED ISSUE #8: [WIP] Change interface between FNS and Worker
Current interface between FNS and Worker needs to be changed.
1. Provide global functions for data reading/writing
    To support mesapy and splited workers, global functions are needed. Then the worker can directly read/write data with context-id/context-token. 
2. A structure for Running Task. 
    The structure will hold the task information, store results and enforce some security policies. 
    The instances of the structure will be saved to the global area. So the above global functions can retrieve the instance and then read/write data. 
3.  A structure for Worker Info and a structure for Worker
4. Worker queue
      The FNS should maintain the available workers

### PR DESCRIPTION
**Pull request by litongxin1991**
## Description

1. Provide global functions for reading/writing files
2. Add a structure to represent running task
3. Add structures to represent worker information and worker
4. Add a queue to maintain available workers
5. Change the way FNS invokes the task 
5. Update current worker functions in FNS 

Fixes #8

## Type of change (select applied and DELETE the others)

- [X] New feature (non-breaking change which adds functionality)
- [X] API change with a documentation update

## Checklist (check ALL before submitting PR, even not applicable)

- [X] Fork the repo and create your branch from `master`.
- [X] If you've added code that should be tested, add tests.
- [X] If you've changed APIs, update the documentation.
- [X] Ensure the tests pass (see CI results).
- [X] Make sure your code lints/format.

### CODE REVIEW

File: docs/how_to_add_your_function.md [hunk 1]
```diff
@@ -1,71 +1,114 @@
...
-
-**Convention**
+## Worker Definition
 
 ```rust
-fn function_name(
-  helper: &mut WorkerHelper,   //crate::trait_defs::WorkerHelper
-  input: WorkerInput,          //crate::trait_defs::WorkerInput
-) -> Result<String>            //mesatee_core::Result
+pub struct YourWorker {}
```
[the remark this answers is not in the document]
Comment (litongxin1991) [reply docs/how_to_add_your_function.md]: fixed

File: docs/how_to_add_your_function.md [hunk 2]
```diff
@@ -1,71 +1,114 @@
...
+pub struct YourWorkerWorkerInput {
```
[the remark this answers is not in the document]
Comment (litongxin1991) [reply docs/how_to_add_your_function.md]: It's removed.

**Comment by litongxin1991** [comment]
The way FNS invokes worker has been changed from static dispatching to dynamic dispatching. The queue will save the trait objects.

**Comment by mssun** [comment]
I'm OK with the PR.

Right now, we need two reviewers' approves. @m4sterchain, could you please help to review this PR? Thanks.

File: mesatee_services/fns/sgx_trusted_lib/src/fns.rs [hunk 3]
```diff
@@ -16,26 +16,45 @@
...
-use fns_proto::{InvokeTaskRequest, InvokeTaskResponse};
-
-use crate::sgx::Executor;
-use crate::trait_defs::TaskExecutor;
-
 pub trait HandleRequest {
     fn handle_request(&self) -> Result<InvokeTaskResponse>;
 }
 
+fn invoker_worker(worker: &mut Worker, request: &InvokeTaskRequest) -> Result<InvokeTaskResponse> {
```
Comment (m4sterchain) [inline mesatee_services/fns/sgx_trusted_lib/src/fns.rs]: invoke_worker?

Comment (litongxin1991) [reply mesatee_services/fns/sgx_trusted_lib/src/fns.rs]: fixed

**Pull request merged**